iml-agent/src/action_plugins/stratagem/action_filesync.rs

```diff
@@ -0,0 +1,214 @@
+                .arg("-S")
+                .arg(src_file)
+                .arg(dest_file)
+                .output();
+
+            let output = output.await?;
+            result.push(FidError {
+                fid: fid.fid.clone(),
+                data: fid.data.clone(),
+                errno: output.status.code().unwrap_or(0) as i16,
+            });
+            tracing::debug!(
+                "dsync exited with {} {} {}",
+                output.status,
+                std::str::from_utf8(&output.stdout)?,
+                std::str::from_utf8(&output.stderr)?
+            );
+        } else {
+            rsync_list.push(Work {
+                src_file: src_file.clone(),
+                dest_file: dest_file.clone(),
+            });
+
+            if rsync_list.len() >= 500 {
+                let mut rsync_res = do_rsync(&rsync_list).await?;
+
+                result.append(&mut rsync_res);
+                rsync_res.clear();
+                rsync_list.clear();
+            }
+        }
+    }
+
+    let mut rsync_res = do_rsync(&rsync_list).await?;
+
+    result.append(&mut rsync_res);
+
+    Ok(result)
+}
+
+async fn restore_fids(
+    llapi: LlapiFid,
+    dest_src: &String,
+    fid_list: Vec<FidItem>,
+) -> Result<Vec<FidError>, ImlAgentError> {
+    let mut result = Vec::new();
+
+    for fid in fid_list {
+        let fid_path = llapi.fid2path(&fid.fid)?;
+        let src_file = format!("{}/{}", llapi.mntpt(), &fid_path);
+        let dest_file = format!("{}/{}", &dest_src, &fid_path);
+        let mut dest_dir = PathBuf::from(&src_file);
+
+        dest_dir.pop();
+        fs::create_dir_all(&dest_dir).await?;
+
+        let md = fs::metadata(&dest_file).await?;
+        /* invoking mpifileutils is slow, so only do it for directories
+         * and big files, otherwise just use rsync
+         */
+        let output;
+        if md.is_dir() || (md.len() > 1024 * 1024 * 1024) {
+            output = Command::new("mpirun")
+                .arg("--hostfile")
+                .arg("/etc/iml/filesync-hostfile")
+                .arg("--allow-run-as-root")
+                .arg("dsync")
+                .arg("-S")
+                .arg(dest_file)
+                .arg(src_file)
+                .output();
+        } else {
+            output = Command::new("rsync").arg(dest_file).arg(src_file).output();
+        }
+
+        let output = output.await?;
+
+        let res = FidError {
+            fid: fid.fid.clone(),
+            data: fid.data.clone(),
+            errno: output.status.code().unwrap_or(0) as i16,
+        };
+        result.push(res);
+
+        /*error!(
+            "exited with {} {} {}",
+            output.status,
+            std::str::from_utf8(&output.stdout)?,
+            std::str::from_utf8(&output.stderr)?
+        );*/
+    }
+
+    Ok(result)
+}
+
+/// Process FIDs
+pub async fn process_fids(
+    (fsname_or_mntpath, task_args, fid_list): (String, HashMap<String, String>, Vec<FidItem>),
```

Comment:
This should also be a struct for `task_args`

Reply:
@utopiabound set the API for this

Reply:
Yeah, I'm not commenting on the API (which doesn't need to change), just that we should deserialize into a struct instead of a `HashMap` (which is equivalent but avoids all the lookups).